### PR TITLE
Refactor `HTML.unescape` in native Crystal

### PR DESCRIPTION
**Pull request by straight-shoota**
The previous implementation of `HTML.unescape` based on `String#gsub(Regex)` which is quite inefficient. Not necessarily the regular expression itself, but it involves a lot of intermediary strings.

This patch introduces a native implementation without regular expression and completely allocation-free (except for the result string of course). As a result, performance improves quite a lot.

This implementation accepts bytes that are invalid UTF-8 and just passes them on. The regex-based implementation would raise an exception on invalid bytes.

The key type of `HTML::SINGLE_CHAR_ENTITIES` and `HTML::DOUBLE_CHAR_ENTITIES` needs to change from `String` to `Bytes` in order to avoid allocating strings for looking up entity names.[^1]

The separation of `SINGLE_CHAR_ENTITIES` and `DOUBLE_CHAR_ENTITIES` is also a bit questionable (there's no space benefit and lookup is less efficient). But I'm leaving that as is because it doesn't matter to the core of the algorithm. It could be refactored in a follow-up.

Depends on #13842 (spec improvements)<del>, thus it's a draft for now.</del>

#13843 is related, but independent.

[^1]: It would be nice if it was possible to do a hash lookup on string keys without allocating a search string. Maybe something similar to `StringPool`. Just a thought for further improvements to stdlib for writing such algorithms.

### CODE REVIEW

**Comment by straight-shoota** [comment]
The wasm32 spec fails. It reproduces locally and I can get some more debug information with `WASMTIME_BACKTRACE_DETAILS=1`:

```
Error: failed to run main module `wasm32_std_spec.wasm`

Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x1fed53 - <unknown>!dlmalloc
           1: 0x200db0 - <unknown>!realloc
           2: 0xa3818 - realloc
                           at /home/johannes/src/crystal-lang/crystal/src/gc/none.cr:21:5
           3: 0xa3804 - __crystal_realloc64
                           at /home/johannes/src/crystal-lang/crystal/src/gc.cr:46:3
           4: 0x15642e - realloc
                           at /home/johannes/src/crystal-lang/crystal/src/pointer.cr:359:18
           5: 0x155a73 - realloc_entries
                           at /home/johannes/src/crystal-lang/crystal/src/hash.cr:910:16
           6: 0x154950 - resize
                           at /home/johannes/src/crystal-lang/crystal/src/hash.cr:589:7
           7: 0x153c3b - upsert
                           at /home/johannes/src/crystal-lang/crystal/src/hash.cr:378:11
           8: 0x15330e - []=
                           at /home/johannes/src/crystal-lang/crystal/src/hash.cr:1050:5
           9: 0x71ece - ~HTML::SINGLE_CHAR_ENTITIES:init
                           at /home/johannes/src/crystal-lang/crystal/src/html/entities.cr:3:26
          10: 0x1da3 - __crystal_main
                           at /home/johannes/src/crystal-lang/crystal/src/html/entities.cr:3:3
          11: 0x1fe43c - main_user_code
                           at /home/johannes/src/crystal-lang/crystal/src/crystal/main.cr:129:5
          12: 0xbbf71 - main
                           at /home/johannes/src/crystal-lang/crystal/src/crystal/main.cr:115:7
          13: 0xbbf37 - main
                           at /home/johannes/src/crystal-lang/crystal/src/crystal/main.cr:141:3
          14: 0xbc029 - __main_argc_argv
                           at /home/johannes/src/crystal-lang/crystal/src/crystal/system/wasi/main.cr:32:3
          15: 0x201946 - <unknown>!__main_void
          16: 0xbbfd8 - _start
                           at /home/johannes/src/crystal-lang/crystal/src/crystal/system/wasi/main.cr:25:12
    2: memory fault at wasm address 0x73657268 in linear memory of size 0x70000
    3: wasm trap: out of bounds memory access
```

Considering that the code works flawlessly in other backends, I presume the error is caused by wasm and not a bug in this patch?

**Comment by HertzDevil** [comment]
What happens if those `Hash`es are preallocated with initial capacities?

**Comment by straight-shoota** [comment]
This is pretty clearly a bug of the wasm32 platform integration.
After transforming the literals, the same error happens in `SINGLE_CHAR_ENTITIES:init`: https://github.com/straight-shoota/crystal/actions/runs/6695871903/job/18192382614
After refactoring that, it now errors in `OptionParser#parse`: https://github.com/straight-shoota/crystal/actions/runs/6696906714/job/18195650780

We need to address that bug separately. It should not hold up this PR.
However we'll need to figure out how to keep going, probably disable HTML specs on wasm32?

**Comment by Fryguy** [comment]
> It would be nice if it was possible to do a hash lookup on string keys without allocating a search string. Maybe something similar to StringPool. Just a thought for further improvements to stdlib for writing such algorithms.

Ruby does something similar internally by duplicating and freezing strings keys on the way in if they aren't already frozen<sup>[[ref]](https://github.com/ruby/ruby/blob/f0b7895637d4c4e4a1c72cecaa8700495f12177b/hash.c#L2838-L2856)</sup>. Frozen strings are pooled in a table and are accessed internally with [rb_fstring and friends](https://github.com/ruby/ruby/blob/master/string.c#L406-L465).

```ruby
> x = "test"
> x.object_id
=> 380
> h = {x => 1}
> h.keys.first.object_id
=> 400
> h.keys.first.frozen?
=> true
> h2 = {x => 2}
> h2.keys.first.object_id
=> 400
> h3 = {"test" => 3}
> h3.keys.first.object_id
=> 400
> x2 = "test".freeze
> x2.object_id
=> 400
> h4 = {x2 => 4}
> h4.keys.first.object_id
=> 400
```

**Review by HertzDevil — CHANGES_REQUESTED** [review]

File: src/html.cr [hunk 1]
```diff
@@ -114,47 +114,77 @@ module HTML
...
+    end
+
+    size = ptr - start_ptr
+    if size > 8
+      # size exceeds maxlength, so it can't be a valid codepoint and might have
+      # overflow.
+      x = 0_u32
+    end
+
+    unless size > 2 && (char = decode_codepoint(x))
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/html.cr]: I don't understand what you mean in the first sentence. Could you elaborate please?

File: src/html.cr [hunk 2]
```diff
@@ -114,47 +114,83 @@ module HTML
...
+    end
+
+    if ptr - number_start_ptr > 8
+      # size exceeds maxlength, so it can't be a valid codepoint and might have
+      # overflow.
+      x = 0_u32
+    end
+
+    size = ptr - start_ptr
+    unless size > 2 && (char = decode_codepoint(x))
```
Comment (HertzDevil) [inline src/html.cr:181]: `size == 2` if the input string is `&#9;`, but it should be transformed to `\t`. I don't think `size == 2` is correct here

**Review by HertzDevil — CHANGES_REQUESTED** [review]

File: src/html.cr (hunk 2 above)
Comment (straight-shoota) [reply src/html.cr:181]: Ah, I see. Yeah `size > 2` is only correct for hex digits. I'll fix it.

**Review by HertzDevil — APPROVED** [review]

**Pull request merged**